### PR TITLE
fix(cli): add optional password setting for headless redis client (#19035)

### LINKED ISSUE #19035: `argocd app diff` fails when Redis is password protected
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Similar to the problem resolved by #18951, `argocd app diff` also will not authenticate when Redis is password protected.

**To Reproduce**

Run `argocd app diff` command on any of the patched versions described in this CVE: https://github.com/argoproj/argo-cd/security/advisories/GHSA-9766-5277-j5hr.

It will throw this error:
```
ERRO[0006] finished unary call with code Unknown         error="error getting cached app managed resources: NOAUTH Authentication required." grpc.code=Unknown grpc.method=ManagedResources grpc.service=application.ApplicationService grpc.start_time="2024-07-12T10:07:34-05:00" grpc.time_ms=545.331 span.kind=server system=grpc
FATA[0006] rpc error: code = Unknown desc = error getting cached app managed resources: NOAUTH Authentication required.
```
**Expected behavior**

I expect diff to behave normally. It was working for me before I upgraded to a version that mitigated the aforementioned CVE.

**Version**

```shell
argocd: v2.11.4+e1284e1
  BuildDate: 2024-07-02T23:16:22Z
  GitCommit: e1284e19e03c9abab2ea55314b14b1e0381c4045
  GitTreeState: clean
  GoVersion: go1.22.4
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.11.4+e1284e1
  BuildDate: 2024-07-02T23:16:22Z
  GitCommit: e1284e19e03c9abab2ea55314b14b1e0381c4045
  GitTreeState: clean
  GoVersion: go1.22.4
  Compiler: gc
  Platform: darwin/arm64
  Kustomize Version: could not get kustomize version: exec: "kustomize": executable file not found in $PATH
  Helm Version: v3.13.3+gc8b9489
  Kubectl Version: v0.26.11
  Jsonnet Version: v0.20.0
```

**Logs**

```
$ argocd app diff <application name>
ERRO[0006] finished unary call with code Unknown         error="error getting cached app managed resources: NOAUTH Authentication required." grpc.code=Unknown grpc.method=ManagedResources grpc.service=application.ApplicationService grpc.start_time="2024-07-12T10:07:34-05:00" grpc.time_ms=545.331 span.kind=server system=grpc
FATA[0006] rpc error: code = Unknown desc = error getting cached app managed resources: NOAUTH Authentication required.
```

### PR DESCRIPTION
**Pull request by sheikhrachel**
# Purpose
- Fixes typos for & exports `DefaultRedisInitialPasswordSecretName` & `DefaultRedisInitialPasswordKey`
- Updates the invocation of `headless.MaybeStartLocalServer()`'s Redis client to optionally use a defined Redis password in the kube config (re: https://github.com/argoproj/argo-cd/pull/18951 in order to attempt to address https://github.com/argoproj/argo-cd/issues/19035)

Fixes [19035](https://github.com/argoproj/argo-cd/issues/19035)

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: cmd/argocd/commands/admin/redis_initial_password.go [hunk 1]
```diff
@@ -23,8 +23,8 @@ import (
 )
 
 const (
-	defaulRedisInitialPasswordSecretName = "argocd-redis"
-	defaultResisInitialPasswordKey       = "auth"
+	DefaultRedisInitialPasswordSecretName = "argocd-redis"
+	DefaultRedisInitialPasswordKey        = "auth"
```
Comment (crenshaw-dev) [inline cmd/argocd/commands/admin/redis_initial_password.go:27]: This can probably be moved to common/common.go to help avoid import cycled.

Comment (sheikhrachel) [reply cmd/argocd/commands/admin/redis_initial_password.go]: Thanks!  I shifted them in there and consolidated the secret fetching logic as well

**Comment by sheikhrachel** [comment]
Only failing check is the licences check.  Happy to get it passing if someone can point me towards what I need to do to resolve the issues there

File: common/common.go [hunk 2]
```diff
@@ -414,3 +419,27 @@ const TokenVerificationError = "failed to verify the token"
...
+/*
+SetOptionalRedisPasswordFromKubeConfig sets the optional Redis password if it exists in the k8s namespace's secrets.
+
+We specify kubeClient as kubernetes.Interface to allow for mocking in tests, but this should be treated as a kubernetes.Clientset param.
+*/
+func SetOptionalRedisPasswordFromKubeConfig(ctx context.Context, kubeClient kubernetes.Interface, namespace string, redisOptions *redis.Options) error {
+	secret, err := kubeClient.CoreV1().Secrets(namespace).Get(ctx, DefaultRedisInitialPasswordSecretName, v1.GetOptions{})
+	if err != nil || secret == nil {
+		return errors.Wrapf(err, "failed to get secret %s/%s", namespace, DefaultRedisInitialPasswordSecretName)
+	}
```
Comment (crenshaw-dev) [inline common/common.go:438]:
```suggestion
	if err != nil {
		return fmt.Errorf("failed to get secret %s/%s: %w", namespace, DefaultRedisInitialPasswordSecretName, err)
	}
	if secret == nil {
		return fmt.Errorf("failed to get secret %s/%s: secret is nil", namespace, DefaultRedisInitialPasswordSecretName)
	}
```

If `err == nil`, `errors.Wrapf` will return `nil`, so we should handle these two cases differently. https://github.com/pkg/errors/blob/v0.9.1/errors.go#L201

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Needs one last tweak

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**